### ISSUE TITLE
Expose AWS Instance ID as a variable

### ISSUE THREAD

**Issue by h3adache** [hashicorp/packer#7186]
This was discussed in #6894 but the alternative is not feasible for everyone.
Currently we run packer from remote CI/CD runners that do not have ssh access to the VPC and run everything via awscli.
Packer has nice features but the lack of a feature to wait for cloud-init (#2639) and lack of ssh to use the workaround really hurts.
As an alternative, we are using an approach that uses awscli (in shell-local) to check the instance in order to make sure that the cloud-init has completed before shutting down.
This works fine other than that we have to hack a way to find the correct instance/instance id.
It would be nice if the packer ebs viewer exposes the instance id for use in providers

**Comment by rickard-von-essen** [comment]
How do you know cloud-init finished without accessing the instance?

**Comment by h3adache** [comment]
You can access the instance using awscli tools.
Using run command (ssm agent required) you can make the same checks you would have from the shell provisioner without needing ssh.

**Comment by davis190** [comment]
I second this for another similar use case.

Currently we are using a `shell-local` provisioner to upload some scripts and SQL files into S3 prior to building our image. These S3 files need to be available before we can kick off the rest of the build. We are currently using an instance tag to single our scripts that it is alright to continue running.

Process: S3 Upload -> Tag Instance "S3UploadComplete -> Continue with build

In order to tag the instance we have to have the InstanceId and because the InstanceId is not made available we have to query for it. To make sure we are querying for the correct instance we use a variable `"RandomId": "{{uuid}}"` in order to set a tag on the instance called `PackerId` that we can easily query. This is all fine and well except that windows and mac both reference environment variables differently so when we set the tag in `shell-local` we have to have two different `shell-locals` (for different environments). Exposing the environment variable would limit that code.

Mac
```
{
      "type": "shell-local",
      "inline": [
        "AWS_INSTANCE_ID=$(aws ec2 describe-instances --region us-east-1 --filter Name=tag:PackerId,Values=$PackerId --query \"Reservations[].Instances[].InstanceId\" --output text)",
        "echo \"Setting tag on $AWS_INSTANCE_ID\"",
        "aws ec2 create-tags --resources $AWS_INSTANCE_ID --tags Key=Status,Value=S3UploadsComplete --region us-east-1 "
      ],
      "environment_vars": [
        "PackerId={{user `RandomId`}}"
      ],
      "only_on": [
        "linux",
        "darwin"
      ]

    },
```

Windows
```
{
      "type": "shell-local",
      "inline": [
        "$AWS_INSTANCE_ID = (aws ec2 describe-instances --region us-east-1 --filter \"Name=tag:PackerId,Values=$env:PackerId\" --query \"Reservations[].Instances[].InstanceId\")",
        "echo \"Setting tag on $AWS_INSTANCE_ID\"",
        "aws ec2 create-tags --resources $AWS_INSTANCE_ID --tags Key=Status,Value=S3UploadsComplete --region us-east-1 "
      ],
      "environment_vars": [
        "PackerId={{user `RandomId`}}"
      ],
      "only_on": [
        "windows"
      ]

    },
```

If we expose the InstanceId in a variable then I can remove the `only_on` and just run the last command in each in a "commonly used" `shell-local`

**Comment by nywilken** [comment]
Hello this issue was resolved by #7079. The AWS Instance ID is available via the [build template engine](https://www.packer.io/docs/templates/engine/#build) 

You can reference the AWS instance_id  with the `ID` build variable. Example below

```
{
      "type": "shell-local",
      "inline": [
        "echo \"Setting tag on $AWS_INSTANCE_ID\"",
        "aws ec2 create-tags --resources $AWS_INSTANCE_ID --tags Key=Status,Value=S3UploadsComplete --region us-east-1 "
      ],
      "environment_vars": [
        "AWS_INSTANCE_ID={{build `ID`}}"
      ]
    },
```

Please reach out if you run into any issues using the build template engine.
